src/coreclr/jit/codegenarmarch.cpp

```diff
@@ -1879,6 +1879,547 @@ void CodeGen::genCodeForCpBlkHelper(GenTreeBlk* cpBlkNode)
     }
 }
 
+#ifdef TARGET_ARM64
+
+class CountingStream
+{
+public:
+    CountingStream()
+    {
+        instrCount = 0;
+    }
+
+    void LoadPairRegs(int offset, int regSizeBytes)
+    {
+        instrCount++;
+    }
+
+    void StorePairRegs(int offset, int regSizeBytes)
+    {
+        instrCount++;
+    }
+
+    void LoadReg(int offset, int regSizeBytes)
+    {
+        instrCount++;
+    }
+
+    void StoreReg(int offset, int regSizeBytes)
+    {
+        instrCount++;
+    }
+
+    int InstructionCount() const
+    {
+        return instrCount;
+    }
+
+private:
+    int instrCount;
+};
+
+class StoreBlockUnrollHelper
+{
+public:
+    static int AlignUp(int offset, int alignment)
+    {
+        assert(((alignment - 1) & alignment) == 0);
+        return (offset + alignment - 1) & (-alignment);
+    }
+
+    static int GetRegSizeAtLeastBytes(int byteCount)
+    {
+        assert(byteCount > 0);
+        assert(byteCount < 16);
+
+        int regSizeBytes = byteCount;
+
+        if (byteCount > 8)
+        {
+            regSizeBytes = 16;
+        }
+        else if (byteCount > 4)
+        {
+            regSizeBytes = 8;
+        }
+        else if (byteCount > 2)
+        {
+            regSizeBytes = 4;
+        }
+
+        return regSizeBytes;
+    }
+};
+
+class VerifyingStream
+{
+public:
+    VerifyingStream()
+    {
+        canEncodeAllLoads  = true;
+        canEncodeAllStores = true;
+    }
+
+    void LoadPairRegs(int offset, int regSizeBytes)
+    {
+        canEncodeAllLoads = canEncodeAllLoads && CanEncodeLoadOrStorePairRegs(offset, regSizeBytes);
+    }
+
+    void StorePairRegs(int offset, int regSizeBytes)
+    {
+        canEncodeAllStores = canEncodeAllStores && CanEncodeLoadOrStorePairRegs(offset, regSizeBytes);
+    }
+
+    void LoadReg(int offset, int regSizeBytes)
+    {
+        canEncodeAllLoads = canEncodeAllLoads && CanEncodeLoadOrStoreReg(offset, regSizeBytes);
+    }
+
+    void StoreReg(int offset, int regSizeBytes)
+    {
+        canEncodeAllStores = canEncodeAllStores && CanEncodeLoadOrStoreReg(offset, regSizeBytes);
+    }
+
+    bool CanEncodeAllLoads() const
+    {
+        return canEncodeAllLoads;
+    }
+
+    bool CanEncodeAllStores() const
+    {
+        return canEncodeAllStores;
+    }
+
+private:
+    static bool CanEncodeLoadOrStorePairRegs(int offset, int regSizeBytes)
```

Comment:
Typically we encapsulate this encoding-related knowledge in the emitter. E.g., we have `emitter::emitIns_valid_imm_for_ldst_offset()`. Could you use that, and/or add related code there, to avoid putting this knowledge in codegen?